go/vt/vtgate/executor.go

```diff
@@ -1130,9 +1131,26 @@ func (e *Executor) fetchOrCreatePlan(
 	}
 
 	if plan == nil {
-		plan, logStats.CachedPlan, stmt, err = e.getCachedOrBuildPlan(ctx, vcursor, query, bindVars, setVarComment, parameterize, planKey)
-		if err != nil {
-			return nil, nil, nil, err
+		plan, logStats.CachedPlan, stmt, err = e.getCachedOrBuildPlan(ctx, vcursor, query, bindVars, setVarComment, parameterize, planKey, false)
+		if err != nil && preparedPlan && isExecutePath {
+			// The baseline plan failed to build, try to build a optimized plan
+			plan, err = e.tryOptimizedPlan(ctx, vcursor, bindVars, query, setVarComment, parameterize, planKey, plan, err)
+		}
+	}
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if preparedPlan && isExecutePath && !plan.Optimized.Swap(true) {
```

Comment:
Is it not better to `Swap(true)` only if we succeed to optimize the plan?

Reply:
We only want to try planning once.

Reply:
We should check if it is required or not based on the baseline plan type
I will add that logic